src/modules/rc_update/rc_update.cpp

```diff
@@ -451,122 +424,200 @@ RCUpdate::Run()
-			manual_control_setpoint.acro_switch = get_rc_sw2pos_position(rc_channels_s::RC_CHANNELS_FUNCTION_ACRO,
-							      _param_rc_acro_th.get(), _param_rc_acro_th.get() < 0.f);
-			manual_control_setpoint.offboard_switch = get_rc_sw2pos_position(rc_channels_s::RC_CHANNELS_FUNCTION_OFFBOARD,
-					_param_rc_offb_th.get(), _param_rc_offb_th.get() < 0.f);
-			manual_control_setpoint.kill_switch = get_rc_sw2pos_position(rc_channels_s::RC_CHANNELS_FUNCTION_KILLSWITCH,
-							      _param_rc_killswitch_th.get(), _param_rc_killswitch_th.get() < 0.f);
-			manual_control_setpoint.arm_switch = get_rc_sw2pos_position(rc_channels_s::RC_CHANNELS_FUNCTION_ARMSWITCH,
-							     _param_rc_armswitch_th.get(), _param_rc_armswitch_th.get() < 0.f);
-			manual_control_setpoint.transition_switch = get_rc_sw2pos_position(rc_channels_s::RC_CHANNELS_FUNCTION_TRANSITION,
-					_param_rc_trans_th.get(), _param_rc_trans_th.get() < 0.f);
-			manual_control_setpoint.gear_switch = get_rc_sw2pos_position(rc_channels_s::RC_CHANNELS_FUNCTION_GEAR,
-							      _param_rc_gear_th.get(), _param_rc_gear_th.get() < 0.f);
-			manual_control_setpoint.stab_switch = get_rc_sw2pos_position(rc_channels_s::RC_CHANNELS_FUNCTION_STAB,
-							      _param_rc_stab_th.get(), _param_rc_stab_th.get() < 0.f);
-			manual_control_setpoint.man_switch = get_rc_sw2pos_position(rc_channels_s::RC_CHANNELS_FUNCTION_MAN,
-							     _param_rc_man_th.get(), _param_rc_man_th.get() < 0.f);
-
-			/* publish manual_control_setpoint topic */
-			_manual_control_setpoint_pub.publish(manual_control_setpoint);
-
-			/* copy from mapped manual_control_setpoint control to control group 3 */
-			actuator_controls_s actuator_group_3{};
-
-			actuator_group_3.timestamp = input_rc.timestamp_last_signal;
-
-			actuator_group_3.control[0] = manual_control_setpoint.y;
-			actuator_group_3.control[1] = manual_control_setpoint.x;
-			actuator_group_3.control[2] = manual_control_setpoint.r;
-			actuator_group_3.control[3] = manual_control_setpoint.z;
-			actuator_group_3.control[4] = manual_control_setpoint.flaps;
-			actuator_group_3.control[5] = manual_control_setpoint.aux1;
-			actuator_group_3.control[6] = manual_control_setpoint.aux2;
-			actuator_group_3.control[7] = manual_control_setpoint.aux3;
-
-			/* publish actuator_controls_3 topic */
-			_actuator_group_3_pub.publish(actuator_group_3);
+			// limit processing if there's no update
+			if (rc_updated || (hrt_elapsed_time(&_last_manual_control_setpoint_publish) > 300_ms)) {
+				UpdateManualSetpoint(input_rc.timestamp_last_signal);
+			}
+
+			UpdateManualSwitches(input_rc.timestamp_last_signal);
 
 			/* Update parameters from RC Channels (tuning with RC) if activated */
 			if (hrt_elapsed_time(&_last_rc_to_param_map_time) > 1_s) {
 				set_params_from_rc();
 				_last_rc_to_param_map_time = hrt_absolute_time();
 			}
 		}
+
+		memcpy(_rc_values_previous, input_rc.values, sizeof(input_rc.values[0]) * input_rc.channel_count);
```

Comment:
channel_limit is min(input_rc.channel_count, RC_MAX_CHAN_COUNT)

Reply:
Exactly and `_rc_values_previous` is an array of length `RC_MAX_CHAN_COUNT` so by using the limited value you'd prevent memory corruption when I send you a malicious `input_rc` message with `input_rc.channel_count = 999;` because I wrote a buggy RC driver for the new protocol that can go into this corner case if you lose RC in flight.

I would even do
`memcpy(_rc_values_previous, input_rc.values, sizeof(_rc_values_previous[0]) * RC_MAX_CHAN_COUNT);`
or
`memcpy(_rc_values_previous, input_rc.values, sizeof(_rc_values_previous[0]) * channel_limit);`
to be on the safe side with the overwrite size.

Reply:
Ha, I misread this and thought it was already using channel_limit. Yes, what you're suggesting makes total sense, so much so that I thought I had already done it...